internal/app/connectconformance/server_runner.go

```diff
@@ -26,8 +26,145 @@ import (
 	"connectrpc.com/conformance/internal"
 	conformancev1alpha1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1alpha1"
 	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
 )
 
+// populateExpectedRequests populates the requests we expected to get back from the server
+// by examining the requests we sent.
+func populateExpectedRequests(testCase *conformancev1alpha1.TestCase) error {
+	// If an expected response was already provided, return and use that.
+	// This allows for overriding this function with explicit values in the yaml file.
+	if testCase.ExpectedResponse != nil {
+		return nil
+	}
+	if len(testCase.Request.RequestMessages) == 0 {
+		return errors.New("at least one request is required")
+	}
+
+	req := testCase.Request.RequestMessages[0]
+
+	switch testCase.Request.StreamType {
+	case conformancev1alpha1.StreamType_STREAM_TYPE_FULL_DUPLEX_BIDI_STREAM,
+		conformancev1alpha1.StreamType_STREAM_TYPE_HALF_DUPLEX_BIDI_STREAM,
+		conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM:
+
+		// First, find the response definition that the client instructed the
+		// server to return
+		var def *conformancev1alpha1.StreamResponseDefinition
+		if testCase.Request.StreamType == conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM {
+			ur := &conformancev1alpha1.ServerStreamRequest{}
+			err := req.UnmarshalTo(ur)
+			if err != nil {
+				return err
+			}
+			def = ur.ResponseDefinition
+		} else {
+			ur := &conformancev1alpha1.BidiStreamRequest{}
+			err := req.UnmarshalTo(ur)
+			if err != nil {
+				return err
+			}
+			def = ur.ResponseDefinition
+		}
+
+		// Server should have echoed back all specified headers, trailers, and errors
+		expected := &conformancev1alpha1.ClientResponseResult{
+			ResponseHeaders:  def.ResponseHeaders,
+			ResponseTrailers: def.ResponseTrailers,
+			Error:            def.Error,
+		}
+
+		// There should be one payload for every ResponseData the client specified
+		expected.Payloads = make([]*conformancev1alpha1.ConformancePayload, len(def.ResponseData))
+
+		for idx, data := range def.ResponseData {
+			expected.Payloads[idx] = &conformancev1alpha1.ConformancePayload{
+				Data: data,
+			}
+			switch testCase.Request.StreamType { //nolint:exhaustive
+			case conformancev1alpha1.StreamType_STREAM_TYPE_SERVER_STREAM:
+				// For server streams, only one request was sent. All request information specified
+				// should only be echoed back in the first response
+				if idx == 0 {
+					expected.Payloads[idx].RequestInfo = &conformancev1alpha1.ConformancePayload_RequestInfo{
+						RequestHeaders: testCase.Request.RequestHeaders,
+						Requests:       testCase.Request.RequestMessages,
+					}
+				}
+			case conformancev1alpha1.StreamType_STREAM_TYPE_HALF_DUPLEX_BIDI_STREAM:
+				// For half-duplex streams, all request information should be echoed back in all
```

Comment:
> all request information should be echoed back in all

Hmmm. I didn't realize it was implemented this way. There's no reason to assert this information repeatedly. It should only be in the first response. There's even a bit of logic in the test runner, that checks for this, that only care about the first message:
https://github.com/connectrpc/conformance/blob/v2/internal/app/connectconformance/results.go#L338

(That's for things like headers and query string params, since they can never change. The request messages can change from one to the next, which is expected to be the case for full-duplex bidi streams.)

Reply:
Changing this means you can collapse this into the above block, since the logic would then be the same as for server streams.

Reply:
Fixed and I also updated the server impls to handle this better. Now, the request information (more accurately -- anything that doesn't change from request to request) is only echoed back in the first response.